cql3/restrictions/statement_restrictions.cc

```diff
@@ -1068,6 +1068,110 @@ std::vector<query::clustering_range> get_single_column_clustering_bounds(
+
+            for (auto&& value : list) {
+                ck_ranges.emplace_back(query::clustering_range::make_singular(std::vector<managed_bytes>{value}));
+            }
+
+            return ck_ranges;
+        },
+        [](const nonwrapping_interval<managed_bytes>& range) {
+            auto int64_from_be_bytes = [](const managed_bytes& int_bytes) -> int64_t {
+                if (int_bytes.size() != 8) {
+                    throw std::runtime_error("token restriction value should be 8 bytes");
+                }
+
+                return read_be<int64_t>((const char*)&int_bytes[0]);
+            };
+
+            auto int64_to_be_bytes = [](int64_t int_val) -> managed_bytes {
+                managed_bytes int_bytes(managed_bytes::initialized_later{}, 8);
+                write_be((char*)&int_bytes[0], int_val);
+                return int_bytes;
+            };
+
+            int64_t token_low = std::numeric_limits<int64_t>::min();
+            int64_t token_high = std::numeric_limits<int64_t>::max();
+            bool low_inclusive = true, high_inclusive = true;
+
+            const std::optional<interval_bound<managed_bytes>>& start = range.start();
+            const std::optional<interval_bound<managed_bytes>>& end = range.end();
+
+            if (start.has_value()) {
+                token_low = int64_from_be_bytes(start->value());
+                low_inclusive = start->is_inclusive();
+            }
+
+            if (end.has_value()) {
+                token_high = int64_from_be_bytes(end->value());
+                high_inclusive = end->is_inclusive();
+            }
+
+            query::clustering_range::bound lower_bound(std::vector{int64_to_be_bytes(token_low)}, low_inclusive);
+            query::clustering_range::bound upper_bound(std::vector{int64_to_be_bytes(token_high)}, high_inclusive);
+
+            std::vector<query::clustering_range> ck_ranges;
+
+            // Blob encoded tokens are sorted like this:
+            // 0, 1, 2, 3, 4, 5, ..., bigint_max, bigint_min, ...., -5, -4, -3, -2, -1
+            // This means that in cases where low >= 0 or high < 0 we can simply use the whole range.
+            // In other cases we have to take two ranges: (low, -1) and (0, high).
+            if (token_low >= 0 || token_high < 0) {
+                ck_ranges.emplace_back(std::move(lower_bound), std::move(upper_bound));
+            } else {
+                query::clustering_range::bound zero_bound(std::vector{int64_to_be_bytes(0)});
+                query::clustering_range::bound min1_bound(std::vector{int64_to_be_bytes(-1)});
+
+                ck_ranges.reserve(2);
+
+                if (!(token_high == 0 && !high_inclusive)) {
+                    ck_ranges.emplace_back(std::move(zero_bound), std::move(upper_bound));
+                }
+
+                if (!(token_low == -1 && !low_inclusive)) {
+                    ck_ranges.emplace_back(std::move(lower_bound), std::move(min1_bound));
+                }
```

Comment:
You could use 0 in both cases, for the non-negative half with inclusive bound and for the negative half with an exclusive half. It has no practical effect, just more mathematically pleasing (splitting the integer axis at one place instead of 0, 1).

Reply:
Actually no, what am I saying, it's not the integer axis any more.